text/0000-improved-itertools-api.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: improved-itertools-api
+- Start Date: 2017-08-8
+- RFC PR: 
+- Pony Issue: 
+
+# Summary
+
+The goal of this RFC is to improve the usability and reduce the maintenance cost of the itertools package.
+
+# Motivation
+
+The motivation for this RFC comes from concerns that I have gathered from my own experiences and from the concerns and recommendations of members of the Pony community. The primary concern that I have heard is inconvenience caused by `fold` being partial. Also, adapting iterators with any state is impossible since the arguments to methods such as `map` and `filter` are lambdas, which means that their `apply` methods have a capability of box. This lack of stateful adaptations also has the side effect of making the `Iter` class prone to code duplication to handle common tasks like stashing values in order to maintain the semantics of filtering. With the addition of methods like `flat_map` and `filter_map`, the problem of code duplication becomes a more noticeable issue for future additions to the API.
+
+# Detailed design
+
+The fold method will be modified so that its argument may not be partial.
+
+```pony
+fun ref fold[B](f: {(B, A!): B^} box, acc: B): B^
+```
+
+A partial version of the method will be provided as well.
+
+```pony
+fun ref fold_partial[B](f: {(B, A!): B^ ?} box, acc: B): B^ ?
+  """
+  A partial version of `fold`.
+  """
+```
+
+The following "iterator adapter" interfaces will be added to the package along with corresponding methods of the `Iter` class. These interfaces will facilitate stateful modification to the output of the iterator.
+
+```pony
+interface IterMapAdapter[A, B]
+  fun ref apply(a: A!): B ?
```

Comment:
I had no idea that this was possible, I'll likely rethink the adapter interfaces given this possibility.